README.md
# Thinkgo    [![GoDoc](https://godoc.org/github.com/tsuna/gohbase?status.png)](https://godoc.org/github.com/henrylee2cn/thinkgo)

![Lessgo Favicon](https://github.com/henrylee2cn/thinkgo/raw/master/doc/thinkgo_96x96.png)

# 概述
Thinkgo以全新的架构实现，它面向Handler接口开发，支持智能参数映射与校验、支持自动化API文档的Go语言web框架。

官方QQ群：Go-Web 编程 42730308    [![Go-Web 编程群](http://pub.idqqimg.com/wpa/images/group.png)](http://jq.qq.com/?_wv=1027&k=fzi4p1)

![thinkgo server](https://github.com/henrylee2cn/thinkgo/raw/master/doc/server.png)

![thinkgo apidoc](https://github.com/henrylee2cn/thinkgo/raw/master/doc/apidoc.png)

# 框架下载

```sh
go get github.com/henrylee2cn/thinkgo
```

# 安装要求

Go Version ≥1.6

# 最新功能特性

- 面向Handler接口开发（func or struct），中间件与操作完全等同可任意拼接路由操作链
- 支持用struct Handler在Tag标签定义请求参数信息及其校验信息
- 由struct Handler自动构建API文档（swagger2.0）
- 支持HTTP/HTTP2、HTTPS(tls/letsencrypt)、UNIX多种Server类型
- 支持多实例运行，且配置信息相互独立
- 基于著名的httprouter构建路由器，且支持链式与树形两种路由注册风格
- 跨平台的彩色日志系统，且同时支持console和file两种输出形式（可以同时使用）
- 提供Session管理功能
- 支持Gzip全局配置
- 提供XSRF跨站请求伪造安全过滤
- 提供静态文件缓存功能
- 排版漂亮的配置文件，且自动补填默认值方便设置


# 代码示例
```
package main

import (
    "github.com/henrylee2cn/thinkgo"
    "time"
)

type Index struct {
    Id        int      `param:"in(path),required,desc(ID),range(0:10)"`
    Title     string   `param:"in(query),nonzero"`
    Paragraph []string `param:"in(query),name(p),len(1:10)" regexp:"(^[\\w]*$)"`
    Cookie    string   `param:"in(cookie),name(thinkgoID)"`
    // Picture         multipart.FileHeader `param:"in(formData),name(pic),maxmb(30)"`
}

func (i *Index) Serve(ctx *thinkgo.Context) error {
    if ctx.CookieParam("thinkgoID") == "" {
        ctx.SetCookie("thinkgoID", time.Now().String())
    }
    return ctx.JSON(200, i)
}

func main() {
  app := thinkgo.New("myapp", "0.1")

  // Register the route in a chain style
  app.GET("/index/:id", new(Index))

  // Register the route in a tree style
  // app.Route(
  //   app.NewGET("/index/:id", new(Index)),
  // )

  // Start the service
  app.Run()
}

/*
http GET:
    http://localhost:8080/index/1?title=test&p=abc&p=xyz
response:
    {
      "Id": 1,
      "Title": "test",
      "Paragraph": [
        "abc",
        "xyz"
      ],
      "Cookie": "2016-11-13 01:14:40.9038005 +0800 CST"
    }
*/
```

# 配置文件说明

- 应用的各实例均有单独一份配置，其文件名格式 `config/{appname}[_{version}].ini`，配置详情：

```
run_mode               = dev                 # 运行模式 dev | prod
net_type               = normal              # Server类型 normal | tls | letsencrypt | unix
addr                   = 0.0.0.0:8080        # 监控地址
tls_certfile           =                     # TLS证书文件路径
tls_keyfile            =                     # TLS密钥文件路径
letsencrypt_file       =                     # SSL免费证书路径
unix_filemode          = 438                 # UNIX Server的文件权限（438即0666）
read_timeout           = 0                   # 读取请求数据超时
write_timeout          = 0                   # 写入响应数据超时
multipart_maxmemory_mb = 32                  # 接收上传文件时允许使用的最大内存

[router]                                     # 路由配置区
redirect_trailing_slash   = true             # 当前请求含`/`后缀的URL如`/foo/`时，若路由不存在但`/foo`，则自动跳转至`/foo`
redirect_fixed_path       = true             # 自动修复URL，如`/FOO` `/..//Foo`均被跳转至`/foo`（依赖redirect_trailing_slash=true）
handle_method_not_allowed = true             # 若开启，当前请求方法不存在时返回405，否则返回404
handle_options            = true             # 若开启，自动应答OPTIONS类请求，可在Thinkgo中设置默认Handler

[xsrf]                                       # XSRF跨站请求伪造过滤配置区
enable = false                               # 是否开启
key    = thinkgoxsrf                         # 加密key
expire = 3600                                # xsrf防伪token有效时长

[session]                                    # Session配置区（详情参考beego session模块）
enable                 = false               # 是否开启
provider               = memory              # 数据存储方式
name                   = thinkgosessionID    # 客户端存储cookie的名字
gc_max_lifetime        = 3600                # 触发GC的时间
provider_config        =                     # 配置信息，根据不同的引擎设置不同的配置信息
cookie_lifetime        = 0                   # 客户端存储的cookie的时间，默认值是0，即浏览器生命周期
auto_setcookie         = true                # 是否自动设置关于session的cookie值，一般默认true
domain                 =                     # 可以访问此cookie的域名
enable_sid_in_header   = false               # 是否将session ID写入Header
name_in_header         = Thinkgosessionid    # 将session ID写入Header时的头名称
enable_sid_in_urlquery = false               # 是否将session ID写入url的query部分

[apidoc]                                     # API文档
enable      = true                           # 是否启用
path        = /apidoc                        # 访问的URL路径
nolimit     = false                          # 是否不限访问IP
real_ip     = false                          # 使用真实客户端的IP进行过滤
prefix_list = 127.|192.168.                  # 允许访问的IP前缀列表
desc        =                                # 项目描述
email       =                                # 联系人邮箱
terms_url   =                                # 服务条款URL
license     =                                # 协议类型
license_url =                                # 协议内容URL
```

- 应用只有一份全局配置，文件名为 `config/__global__.ini`，配置详情：

```
[cache]                                      # 文件内存缓存配置区
enable  = false                              # 是否开启
size_mb = 32                                 # 允许缓存使用的最大内存（单位MB），为0时系统自动设置为512KB
expire  = 60                                 # 缓存最大时长

[gzip]                                       # gzip压缩配置区
enable         = false                       # 是否开启
min_length     = 20                          # 进行压缩的最小内容长度
compress_level = 1                           # 非文件类响应Body的压缩水平（0-9），注意文件压缩始终为最优压缩比（9）
methods        = GET                         # 允许压缩的请求方法，为空时默认为GET

[log]                                        # 日志配置区
console_enable = true                        # 是否启用控制台日志
console_level  = debug                       # 控制台日志打印水平
file_enable    = true                        # 是否启用文件日志
file_level     = debug                       # 文件日志打印水平
```

# Handler结构体字段标签说明

tag   |   key    | required |     value     |   desc
------|----------|----------|---------------|----------------------------------
param |    in    | only one |     path      | (position of param) if `required` is unsetted, auto set it. e.g. url: "http://www.abc.com/a/{path}"
param |    in    | only one |     query     | (position of param) e.g. url: "http://www.abc.com/a?b={query}"
param |    in    | only one |     formData  | (position of param) e.g. "request body: a=123&b={formData}"
param |    in    | only one |     body      | (position of param) request body can be any content
param |    in    | only one |     header    | (position of param) request header info
param |    in    | only one |     cookie    | (position of param) request cookie info, support: `http.Cookie`, `fasthttp.Cookie`, `string`, `[]byte` and so on
param |   name   |    no    |  (e.g. "id")  | specify request param`s name
param | required |    no    |   required    | request param is required
param |   desc   |    no    |  (e.g. "id")  | request param description
param |   len    |    no    | (e.g. 3:6, 3) | length range of param's value
param |   range  |    no    |  (e.g. 0:10)  | numerical range of param's value
param |  nonzero |    no    |    nonzero    | param`s value can not be zero
param |   maxmb  |    no    |   (e.g. 32)   | when request Content-Type is multipart/form-data, the max memory for body.(multi-param, whichever is greater)
regexp|          |    no    |(e.g. "^\\w+$")| param value can not be null
err   |          |    no    |(e.g. "incorrect password format")| customize the prompt for validation error

**NOTES**:
* the binding object must be a struct pointer
* the binding struct's field can not be a pointer
* `regexp` or `param` tag is only usable when `param:"type(xxx)"` is exist
* if the `param` tag is not exist, anonymous field will be parsed
* when the param's position(`in`) is `formData` and the field's type is `multipart.FileHeader`, the param receives file uploaded
* if param's position(`in`) is `cookie`, field's type must be `http.Cookie`
* param tags `in(formData)` and `in(body)` can not exist at the same time
* there should not be more than one `in(body)` param tag

# Handler结构体字段类型说明

base    |   slice    | special
--------|------------|-------------------------------------------------------
string  |  []string  | [][]byte
byte    |  []byte    | [][]uint8
uint8   |  []uint8   | multipart.FileHeader (only for `formData` param)
bool    |  []bool    | http.Cookie (only for `net/http`'s `cookie` param)
int     |  []int     | fasthttp.Cookie (only for `fasthttp`'s `cookie` param)
int8    |  []int8    | struct (struct type only for `body` param or as an anonymous field to extend params)
int16   |  []int16   |
int32   |  []int32   |
int64   |  []int64   |
uint8   |  []uint8   |
uint16  |  []uint16  |
uint32  |  []uint32  |
uint64  |  []uint64  |
float32 |  []float32 |
float64 |  []float64 |

## 开源协议
Thinkgo 项目采用商业应用友好的 [Apache2.0](https://github.com/henrylee2cn/thinkgo/raw/master/LICENSE) 协议发布。


logging/CHANGELOG.md
# Changelog

## 2.0.0-rc1 (2016-02-11)

Time flies and it has been three years since this package was first released.
There have been a couple of API changes I have wanted to do for some time but
I've tried to maintain backwards compatibility. Some inconsistencies in the
API have started to show, proper vendor support in Go out of the box and
the fact that `go vet` will give warnings -- I have decided to bump the major
version.

* Make eg. `Info` and `Infof` do different things. You want to change all calls
	to `Info` with a string format go to `Infof` etc. In many cases, `go vet` will
	guide you.
* `Id` in `Record` is now called `ID`

## 1.0.0 (2013-02-21)

Initial release


config.go
// Copyright 2016 HenryLee. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package thinkgo

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync/atomic"
	"time"

	"github.com/henrylee2cn/thinkgo/ini"
)

type (
	GlobalConfig struct {
		Cache   CacheConfig `ini:"cache"`
		Gzip    GzipConfig  `ini:"gzip"`
		Log     LogConfig   `ini:"log"`
		warnMsg string      `int:"-"`
	}
	Config struct {
		// RunMode         string      `ini:"run_mode"`         // run mode: dev | prod
		NetType         string      `ini:"net_type"`         // network type: normal | tls | letsencrypt | unix
		Addr            string      `ini:"addr"`             // Service monitoring address
		TLSCertFile     string      `ini:"tls_certfile"`     // for TLS
		TLSKeyFile      string      `ini:"tls_keyfile"`      // for TLS
		LetsencryptFile string      `ini:"letsencrypt_file"` // for Let's Encrypt (free SSL)
		UNIXFileMode    os.FileMode `ini:"unix_filemode"`    // for UNIX listener file mode
		// Maximum duration for reading the full request (including body).
		//
		// This also limits the maximum duration for idle keep-alive
		// connections.
		//
		// By default request read timeout is unlimited.
		ReadTimeout time.Duration `ini:"read_timeout"`
		// Maximum duration for writing the full response (including body).
		//
		// By default response write timeout is unlimited.
		WriteTimeout         time.Duration `ini:"write_timeout"`
		MultipartMaxMemoryMB int64         `ini:"multipart_maxmemory_mb"`
		multipartMaxMemory   int64         `ini:"-"`
		Router               RouterConfig  `ini:"router"`
		XSRF                 XSRFConfig    `ini:"xsrf"`
		Session              SessionConfig `ini:"session"`
		APIdoc               APIdocConfig  `ini:"apidoc"`
	}
	RouterConfig struct {
		// Enables automatic redirection if the current route can't be matched but a
		// handler for the path with (without) the trailing slash exists.
		// For example if /foo/ is requested but a route only exists for /foo, the
		// client is redirected to /foo with http status code 301 for GET requests
		// and 307 for all other request methods.
		RedirectTrailingSlash bool `ini:"redirect_trailing_slash"`
		// If enabled, the router tries to fix the current request path, if no
		// handle is registered for it.
		// First superfluous path elements like ../ or // are removed.
		// Afterwards the router does a case-insensitive lookup of the cleaned path.
		// If a handle can be found for this route, the router makes a redirection
		// to the corrected path with status code 301 for GET requests and 307 for
		// all other request methods.
		// For example /FOO and /..//Foo could be redirected to /foo.
		// RedirectTrailingSlash is independent of this option.
		RedirectFixedPath bool `ini:"redirect_fixed_path"`
		// If enabled, the router checks if another method is allowed for the
		// current route, if the current request can not be routed.
		// If this is the case, the request is answered with 'Method Not Allowed'
		// and HTTP status code 405.
		// If no other Method is allowed, the request is delegated to the NotFound
		// handler.
		HandleMethodNotAllowed bool `ini:"handle_method_not_allowed"`
		// If enabled, the router automatically replies to OPTIONS requests.
		// Custom OPTIONS handlers take priority over automatic replies.
		HandleOPTIONS bool `ini:"handle_options"`
	}
	GzipConfig struct {
		// if EnableGzip, compress response content.
		Enable bool `ini:"enable"`
		//Content will only be compressed if content length is either unknown or greater than gzipMinLength.
		//Default size==20B same as nginx
		MinLength int `ini:"min_length"`
		//The compression level used for deflate compression. (0-9).
		CompressLevel int `ini:"compress_level"`
		//List of HTTP methods to compress. If not set, only GET requests are compressed.
		Methods []string `ini:"methods" delim:"|"`
		// StaticExtensionsToGzip []string
	}
	CacheConfig struct {
		// Whether to enable caching static files
		Enable bool `ini:"enable"`
		// Max size by MB for file cache.
		// The cache size will be set to 512KB at minimum.
		// If the size is set relatively large, you should call
		// `debug.SetGCPercent()`, set it to a much smaller value
		// to limit the memory consumption and GC pause time.
		SizeMB int64 `ini:"size_mb"`
		// expire in xxx seconds for file cache.
		// Expire <= 0 (second) means no expire, but it can be evicted when cache is full.
		Expire int `ini:"expire"`
	}
	XSRFConfig struct {
		Enable bool   `ini:"enable"`
		Key    string `ini:"key"`
		Expire int    `ini:"expire"`
	}
	SessionConfig struct {
		Enable                bool   `ini:"enable"`
		Provider              string `ini:"provider"`
		Name                  string `ini:"name"`
		GCMaxLifetime         int64  `ini:"gc_max_lifetime"`
		ProviderConfig        string `ini:"provider_config"`
		CookieLifetime        int    `ini:"cookie_lifetime"`
		AutoSetCookie         bool   `ini:"auto_setcookie"`
		Domain                string `ini:"domain"`
		EnableSidInHttpHeader bool   `ini:"enable_sid_in_header"` // enable store/get the sessionId into/from http headers
		NameInHttpHeader      string `ini:"name_in_header"`
		EnableSidInUrlQuery   bool   `ini:"enable_sid_in_urlquery"` // enable get the sessionId from Url Query params
	}
	LogConfig struct {
		ConsoleEnable bool   `ini:"console_enable"`
		ConsoleLevel  string `ini:"console_level"` // critical | error | warning | notice | info | debug
		FileEnable    bool   `ini:"file_enable"`
		FileLevel     string `ini:"file_level"` // critical | error | warning | notice | info | debug
	}
	APIdocConfig struct {
		Enable     bool     `ini:"enable"`                // Whether to enable API doc
		Path       string   `ini:"path"`                  // API doc url
		NoLimit    bool     `ini:"nolimit"`               // if true, access is not restricted
		RealIP     bool     `ini:"real_ip"`               // if true, means verifying the real IP of the visitor
		PrefixList []string `ini:"prefix_list" delim:"|"` // `prefix_list=192;202` means that only IP addresses with 192 or 202 prefixes are allowed
		Desc       string   `ini:"desc"`                  // description of the application
		Email      string   `ini:"email"`                 // technician's Email
		TermsURL   string   `ini:"terms_url"`             // terms of service
		License    string   `ini:"license"`               // the license used by the API
		LicenseURL string   `ini:"license_url"`
	}
)

const (
	// RUNMODE_DEV                 = "dev"
	// RUNMODE_PROD                = "prod"
	NETTYPE_NORMAL              = "normal"
	NETTYPE_TLS                 = "tls"
	NETTYPE_LETSENCRYPT         = "letsencrypt"
	NETTYPE_UNIX                = "unix"
	MB                          = 1 << 20 // 1MB
	defaultMultipartMaxMemory   = 32 * MB // 32 MB
	defaultMultipartMaxMemoryMB = 32
	defaultPort                 = 8080

	// The path for the configuration files
	CONFIG_DIR = "./config/"
	// global config file name
	GLOBAL_CONFIG_FILE = "__global___.ini"
)

var (
	appCount uint32
)

// global config
var globalConfig = func() GlobalConfig {
	var background = GlobalConfig{
		Cache: CacheConfig{
			Enable: false,
			SizeMB: 32,
			Expire: 60,
		},
		Gzip: GzipConfig{
			Enable:        false,
			MinLength:     20,
			CompressLevel: 1,
			Methods:       []string{"GET"},
		},
		Log: LogConfig{
			ConsoleEnable: true,
			ConsoleLevel:  "debug",
			FileEnable:    true,
			FileLevel:     "debug",
		},
	}
	filename := CONFIG_DIR + GLOBAL_CONFIG_FILE
	os.MkdirAll(filepath.Dir(filename), 0777)

	cfg, err := ini.LooseLoad(filename)
	if err != nil {
		panic(err)
	}
	err = cfg.MapTo(&background)
	if err != nil {
		panic(err)
	}

	{
		if !(background.Log.ConsoleEnable || background.Log.FileEnable) {
			background.Log.ConsoleEnable = true
			background.warnMsg = "config: log::enable_console and log::enable_file can not be disabled at the same time, so automatically open console log."
		}
	}

	err = cfg.ReflectFrom(&background)
	if err != nil {
		panic(err)
	}
	err = cfg.SaveTo(filename)
	if err != nil {
		panic(err)
	}
	return background
}()

func newConfig(filename string, addrs ...string) Config {
	var addr string
	if len(addrs) > 0 {
		addr = addrs[0]
	} else {
		addr = fmt.Sprintf("0.0.0.0:%d", defaultPort+atomic.LoadUint32(&appCount))
		atomic.AddUint32(&appCount, 1)
	}
	var background = Config{
		// RunMode:              RUNMODE_DEV,
		NetType:              "normal",
		Addr:                 addr,
		UNIXFileMode:         0666,
		MultipartMaxMemoryMB: defaultMultipartMaxMemoryMB,
		Router: RouterConfig{
			RedirectTrailingSlash:  true,
			RedirectFixedPath:      true,
			HandleMethodNotAllowed: true,
			HandleOPTIONS:          true,
		},
		XSRF: XSRFConfig{
			Enable: false,
			Key:    "thinkgoxsrf",
			Expire: 3600,
		},
		Session: SessionConfig{
			Enable:                false,
			Provider:              "memory",
			Name:                  "thinkgosessionID",
			GCMaxLifetime:         3600,
			ProviderConfig:        "",
			CookieLifetime:        0, //set cookie default is the browser life
			AutoSetCookie:         true,
			Domain:                "",
			EnableSidInHttpHeader: false, //	enable store/get the sessionId into/from http headers
			NameInHttpHeader:      "Thinkgosessionid",
			EnableSidInUrlQuery:   false, //	enable get the sessionId from Url Query params
		},
		APIdoc: APIdocConfig{
			Enable:  true,
			Path:    "/apidoc",
			NoLimit: false,
			RealIP:  false,
			PrefixList: []string{
				"127.",
				"192.168.",
			},
		},
	}

	os.MkdirAll(filepath.Dir(filename), 0777)

	cfg, err := ini.LooseLoad(filename)
	if err != nil {
		panic(err)
	}
	err = cfg.MapTo(&background)
	if err != nil {
		panic(err)
	}

	{
		// switch background.RunMode {
		// case RUNMODE_DEV, RUNMODE_PROD:
		// default:
		// 	panic("Please set a valid config item run_mode, refer to the following:\ndev | prod")
		// }
		switch background.NetType {
		case NETTYPE_NORMAL, NETTYPE_TLS, NETTYPE_LETSENCRYPT, NETTYPE_UNIX:
		default:
			panic("Please set a valid config item net_type, refer to the following:\nnormal | tls | letsencrypt | unix")
		}
		background.multipartMaxMemory = background.MultipartMaxMemoryMB * MB
		background.APIdoc.Comb()
	}

	err = cfg.ReflectFrom(&background)
	if err != nil {
		panic(err)
	}
	err = cfg.SaveTo(filename)
	if err != nil {
		panic(err)
	}
	return background
}

func syncConfigToFile(filename string, config *Config) error {
	cfg, err := ini.LooseLoad(filename)
	if err != nil {
		return err
	}
	err = cfg.ReflectFrom(&config)
	if err != nil {
		return err
	}
	return cfg.SaveTo(filename)
}

func (conf *APIdocConfig) Comb() {
	ipPrefixMap := map[string]bool{}
	for _, ipPrefix := range conf.PrefixList {
		if len(ipPrefix) > 0 {
			ipPrefixMap[ipPrefix] = true
		}
	}
	conf.PrefixList = conf.PrefixList[:0]
	for ipPrefix := range ipPrefixMap {
		conf.PrefixList = append(conf.PrefixList, ipPrefix)
	}
	sort.Strings(conf.PrefixList)
}
